pkg/service/service_positions.go

```diff
@@ -0,0 +1,92 @@
+package service
+
+import (
+	"encoding/json"
+
+	"github.com/readium/go-toolkit/pkg/fetcher"
+	"github.com/readium/go-toolkit/pkg/manifest"
+)
+
+var PositionsLink = manifest.Link{
+	Href: "/~readium/positions.json",
+	Type: "application/vnd.readium.position-list+json",
+}
+
+// PositionsService implements Service
+// Provides a list of discrete locations in the publication, no matter what the original format is.
+type PositionsService interface {
+	Service
+	PositionsByReadingOrder() [][]manifest.Locator // Returns the list of all the positions in the publication, grouped by the resource reading order index.
+	Positions() []manifest.Locator                 // Returns the list of all the positions in the publication. (flattening of PositionsByReadingOrder)
+}
+
+// PerResourcePositionsService implements PositionsService
+// Simple [PositionsService] which generates one position per [readingOrder] resource.
+type PerResourcePositionsService struct {
+	readingOrder      []manifest.Link
+	fallbackMediaType string
+}
+
+func GetForPositionsService(service PositionsService, link manifest.Link) (fetcher.Resource, bool) {
+	if link.Href != PositionsLink.Href {
+		return nil, false
+	}
+
+	return fetcher.NewBytesResource(PositionsLink, func() []byte {
+		positions := service.Positions()
+		bin, _ := json.Marshal(map[string]interface{}{
+			"total":     len(positions),
+			"positions": positions,
+		})
+		return bin
+	}), true
+}
+
+func (s PerResourcePositionsService) Close() {}
+
+func (s PerResourcePositionsService) Links() []manifest.Link {
+	return []manifest.Link{PositionsLink}
+}
+
+func (s PerResourcePositionsService) Get(link manifest.Link) (fetcher.Resource, bool) {
+	return GetForPositionsService(s, link)
+}
+
+func (s PerResourcePositionsService) Positions() []manifest.Locator {
+	poss := s.PositionsByReadingOrder()
+	positions := make([]manifest.Locator, len(poss))
+	for i, v := range poss {
+		positions[i] = v[0] // Always just one element
+	}
+	return positions
+}
+
+func (s PerResourcePositionsService) PositionsByReadingOrder() [][]manifest.Locator {
```

Comment:
We're missing helpers for these two APIs on the `Publication` object. Since we will move the positions service into the `pub` package, you should be able to add them easily.

Reply:
Can you elaborate on what this would look like? Are you saying a function that wraps `publication.Get(Link{Href: "/~readium/positions.json"})` ? Or more of a function that specifically finds the positions service, and specifically calls those functions?

Reply:
Sure, you can find it there: https://github.com/readium/kotlin-toolkit/blob/540c80fc1c57f04d09db45764f103724177aedf9/readium/shared/src/main/java/org/readium/r2/shared/publication/services/PositionsService.kt#L57-L85

I don't think we'll need the `positionsFromManifest` for the Go toolkit though, so it might just look like this:

```go
func (p *Publication) PositionsByReadingOrder() [][]manifest.Locator {
    service := p.FindService(PositionsService)
    if service == nil {
        return nil
    }
    return service.PositionsByReadingOrder()
}

func (p *Publication) Positions() [][]manifest.Locator {
    service := p.FindService(PositionsService)
    if service == nil {
        return nil
    }
    return service.Positions()
}
```

This way the caller can do this directly, without caring (or knowing about) the positions service:

```go
publication := Parse(epubFile)
publication.Positions()
```